### PR TITLE
feat: E2EE.

### PR DESCRIPTION
**Pull request by cloudwebrtc**
Frame Cryptor support for Android.

- [X] E2EE implementation code
- [x] Test it in sample app.

### CODE REVIEW

File: livekit-android-sdk/src/main/java/io/livekit/android/e2ee/KeyProvider.kt [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+     * @param participantId
+     * @param keyIndex
+     */
+    override fun setKey(key: String, participantId: String?, keyIndex: Int?) {
+        if (enableSharedKey) {
+            sharedKey = key.toByteArray()
+            return
+        }
+
+        var keyInfo = KeyInfo("", keyIndex ?: 0, key)
```
Comment (davidliu) [inline livekit-android-sdk/src/main/java/io/livekit/android/e2ee/KeyProvider.kt:46]: Is this intentionally `""` or should the participantId be passed in here?

File: livekit-android-sdk/src/main/java/io/livekit/android/e2ee/KeyProvider.kt [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+        }
+        keys[keyInfo.participantId]!![keyInfo.keyIndex] = keyInfo.key
+        rtcKeyProvider.setKey(participantId, keyIndex?: 0, key.toByteArray())
+    }
+
+    override fun ratchetKey(participantId: String, index: Int): ByteArray {
+        return rtcKeyProvider.ratchetKey(participantId, index)
+    }
+
+    override lateinit var rtcKeyProvider: FrameCryptorKeyProvider
```
Comment (davidliu) [inline livekit-android-sdk/src/main/java/io/livekit/android/e2ee/KeyProvider.kt:59]:
```suggestion
    override val rtcKeyProvider: FrameCryptorKeyProvider
```

File: livekit-android-sdk/src/main/java/io/livekit/android/e2ee/KeyProvider.kt [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+
+    val rtcKeyProvider: FrameCryptorKeyProvider
+
+     var sharedKey: ByteArray?
+
+     var enableSharedKey: Boolean
+}
+
+class BaseKeyProvider
+constructor(ratchetSalt: String, uncryptedMagicBytes: String, ratchetWindowSize: Int, enableSharedKey: Boolean) :
```
Comment (davidliu) [inline livekit-android-sdk/src/main/java/io/livekit/android/e2ee/KeyProvider.kt:25]:
```suggestion
constructor(private val ratchetSalt: String, private val uncryptedMagicBytes: String, private val ratchetWindowSize: Int, private val enableSharedKey: Boolean) :
```

File: livekit-android-sdk/src/main/java/io/livekit/android/room/Room.kt [hunk 4]
```diff
@@ -243,7 +249,17 @@ constructor(
 
         state = State.CONNECTING
         connectOptions = options
-        engine.join(url, token, options, getCurrentRoomOptions())
+
+        if(roomOptions.e2eeOptions != null) {
+            e2eeManager = E2EEManager(roomOptions!!.e2eeOptions!!.keyProvider)
+            e2eeManager!!.setup(this, {event ->
+                CoroutineScope(ioDispatcher).launch {
```
Comment (davidliu) [inline livekit-android-sdk/src/main/java/io/livekit/android/room/Room.kt:256]:
```suggestion
                coroutineScope.launch {
```

File: livekit-android-sdk/src/main/java/io/livekit/android/room/Room.kt [hunk 5]
```diff
@@ -888,6 +906,16 @@ constructor(
         eventBus.postEvent(RoomEvent.TrackUnpublished(this, publication, participant), coroutineScope)
     }
 
+    /**
+     * @suppress
+     */
+    override fun onTrackPublished(publication: LocalTrackPublication, participant: LocalParticipant) {
```
Comment (davidliu) [inline livekit-android-sdk/src/main/java/io/livekit/android/room/Room.kt:912]: This is already handled elsewhere, I think we can remove this.

File: livekit-android-sdk/src/main/java/io/livekit/android/room/track/LocalAudioTrack.kt [hunk 6]
```diff
@@ -26,7 +26,7 @@ class LocalAudioTrack(
         }
 
     internal var transceiver: RtpTransceiver? = null
-    private val sender: RtpSender?
+    val sender: RtpSender?
```
Comment (davidliu) [inline livekit-android-sdk/src/main/java/io/livekit/android/room/track/LocalAudioTrack.kt:29]:
```suggestion
    internal val sender: RtpSender?
```

File: livekit-android-sdk/src/main/java/io/livekit/android/room/track/LocalVideoTrack.kt [hunk 7]
```diff
@@ -61,7 +61,7 @@ constructor(
         }
 
     internal var transceiver: RtpTransceiver? = null
-    private val sender: RtpSender?
+    val sender: RtpSender?
```
Comment (davidliu) [inline livekit-android-sdk/src/main/java/io/livekit/android/room/track/LocalVideoTrack.kt:64]:
```suggestion
    internal val sender: RtpSender?
```

File: livekit-android-sdk/src/main/java/io/livekit/android/room/track/RemoteAudioTrack.kt [hunk 8]
```diff
@@ -1,5 +1,14 @@
 package io.livekit.android.room.track
 
 import org.webrtc.AudioTrack
+import org.webrtc.RtpReceiver
+import org.webrtc.RtpSender
+import org.webrtc.RtpTransceiver
 
-class RemoteAudioTrack(name: String, rtcTrack: AudioTrack) : io.livekit.android.room.track.AudioTrack(name, rtcTrack)
+class RemoteAudioTrack(name: String, rtcTrack: AudioTrack, receiver: RtpReceiver) : io.livekit.android.room.track.AudioTrack(name, rtcTrack)
```
Comment (davidliu) [inline livekit-android-sdk/src/main/java/io/livekit/android/room/track/RemoteAudioTrack.kt:8]:
```suggestion
class RemoteAudioTrack(name: String, rtcTrack: AudioTrack, internal val receiver: RtpReceiver) : io.livekit.android.room.track.AudioTrack(name, rtcTrack)
```

File: livekit-android-sdk/src/main/java/io/livekit/android/room/track/RemoteVideoTrack.kt [hunk 9]
```diff
@@ -18,6 +20,7 @@ class RemoteVideoTrack(
     val autoManageVideo: Boolean = false,
     @Named(InjectionNames.DISPATCHER_DEFAULT)
     private val dispatcher: CoroutineDispatcher,
+    receiver: RtpReceiver,
```
Comment (davidliu) [inline livekit-android-sdk/src/main/java/io/livekit/android/room/track/RemoteVideoTrack.kt:23]:
```suggestion
    internal val receiver: RtpReceiver,
```

File: livekit-android-sdk/src/main/java/io/livekit/android/room/track/RemoteVideoTrack.kt [hunk 10]
```diff
@@ -29,6 +32,12 @@ class RemoteVideoTrack(
     internal var lastDimensions: Dimensions = Dimensions(0, 0)
         private set
 
+    internal var receiver: RtpReceiver
```
Comment (davidliu) [inline livekit-android-sdk/src/main/java/io/livekit/android/room/track/RemoteVideoTrack.kt:35]: remove

File: livekit-android-sdk/src/main/java/io/livekit/android/room/track/RemoteAudioTrack.kt [hunk 11]
```diff
@@ -1,5 +1,14 @@
...
+{
+    internal var receiver: RtpReceiver
```
Comment (davidliu) [inline livekit-android-sdk/src/main/java/io/livekit/android/room/track/RemoteAudioTrack.kt:10]: remove

File: sample-app-common/src/main/java/io/livekit/android/sample/CallViewModel.kt [hunk 12]
```diff
@@ -37,13 +38,32 @@ import kotlinx.coroutines.launch
 class CallViewModel(
     val url: String,
     val token: String,
-    application: Application
+    application: Application,
+    e2ee: Boolean? = false,
```
Comment (davidliu) [inline sample-app-common/src/main/java/io/livekit/android/sample/CallViewModel.kt:42]:
```suggestion
    val e2ee: Boolean = false,
```

File: sample-app-common/src/main/java/io/livekit/android/sample/CallViewModel.kt [hunk 13]
```diff
@@ -37,13 +38,32 @@ import kotlinx.coroutines.launch
...
+    e2eeKey: String? = "",
```
Comment (davidliu) [inline sample-app-common/src/main/java/io/livekit/android/sample/CallViewModel.kt:43]:
```suggestion
    val e2eeKey: String = "",
```

File: livekit-android-sdk/src/main/java/io/livekit/android/room/Room.kt [hunk 14]
```diff
@@ -185,7 +191,7 @@ constructor(
             videoTrackPublishDefaults = videoTrackPublishDefaults,
```
Comment (davidliu) [inline livekit-android-sdk/src/main/java/io/livekit/android/room/Room.kt:191]: need to capture e2eeOptions in the Room (also in LiveKit.kt)

File: livekit-android-sdk/src/main/java/io/livekit/android/e2ee/E2EEManager.kt [hunk 15]
```diff
@@ -0,0 +1,182 @@
...
+    private var keyProvider: KeyProvider
+    private var frameCryptors = mutableMapOf<String, FrameCryptor>()
+    private var algorithm: FrameCryptorAlgorithm = FrameCryptorAlgorithm.AES_GCM
+    private lateinit var emitEvent: (roomEvent: RoomEvent) -> Unit?
+    var enabled: Boolean = false
+    init {
+        this.keyProvider = keyProvider
+    }
+
+    suspend fun setup(room: Room, emitEvent: (roomEvent: RoomEvent) -> Unit) {
```
Comment (davidliu) [inline livekit-android-sdk/src/main/java/io/livekit/android/e2ee/E2EEManager.kt:32]: Looks like these can be put in the constructor?

**Review by davidliu — CHANGES_REQUESTED** [review]

File: livekit-android-sdk/src/main/java/io/livekit/android/e2ee/E2EEManager.kt (hunk 15 above)
Comment (cloudwebrtc) [reply livekit-android-sdk/src/main/java/io/livekit/android/e2ee/E2EEManager.kt:32]: It was discussed with Lukas here that we can enter a certain room without enabling e2ee, and use setup to switch to encrypted mode halfway, so I suggest to keep

**Review by navin-kumar-sheoran — APPROVED** [review]

**Pull request merged**